osquery/tables/ipmi/ipmi_info.cpp

```diff
@@ -0,0 +1,1297 @@
+              return;
+            }
+
+            if (medium == IPMI_CHANNEL_MEDIUM_8023_LAN) {
+              IPMIClient* c = reinterpret_cast<IPMIClient*>(data);
+              c->setLANCh(ch);
+            }
+
+          },
+          data);
+      if (rv != 0) {
+        LOG(ERROR) << "Could not get channel info for channel "
+                   << std::to_string(i) << ": " << strerror(rv);
+      }
+    }
+  });
+  blkAndOp([]() { return false; }, 500);
+}
+
+int IPMIClient::oneOp() {
+  struct timeval tv;
+  tv.tv_sec = 0;
+  tv.tv_usec = 500000;
+
+  return os_hnd_.get()->perform_one_op(os_hnd_.get(), &tv);
+}
+
+void IPMIClient::setDomain(ipmi_domain_t* d) {
+  domain_ = d;
+}
+
+void IPMIClient::iterateEntities(ipmi_entities_iterate_entity_cb cb) {
+  if (domain_ == nullptr) {
+    LOG(ERROR) << "No IPMI domain pointer..nothing to do";
+    return;
+  }
+
+  int rv = ipmi_domain_iterate_entities(domain_, cb, this);
+  if (rv != 0) {
+    LOG(ERROR) << "Could not register callback for "
+                  "ipmi_domain_iterate_entities: "
+               << strerror(rv);
+    return;
+  }
+}
+
+void IPMIClient::iterateMCs(ipmi_domain_iterate_mcs_cb cb) {
+  if (domain_ == nullptr) {
+    LOG(ERROR) << "No IPMI domain pointer..nothing to do";
+    return;
+  }
+
+  int rv = ipmi_domain_iterate_mcs(domain_, cb, this);
+  if (rv != 0) {
+    LOG(ERROR) << "Could not register callback for ipmi_domain_iterate_mcs: "
+               << strerror(rv);
+    return;
+  }
+}
+
+void IPMIClient::getLANConfigs(QueryData& results) {
+  WriteLock lock(busy_);
+
+  rowsQueue_.clear();
+
+  iterateMCs(getLANsCB);
+  blkAndOp([]() { return false; }, 500);
+  toQueryData(results);
+
+  rowsQueue_.clear();
+}
+
+void IPMIClient::getThresholdSensors(QueryData& results) {
+  WriteLock lock(busy_);
+  rowsQueue_.clear();
+
+  iterateEntities(getThresholdSensorCB);
+  blkAndOp([]() { return false; }, 500);
+  toQueryData(results);
+
+  rowsQueue_.clear();
+}
+
+void IPMIClient::getFRUs(QueryData& results) {
+  WriteLock lock(busy_);
+  rowsQueue_.clear();
+
+  iterateEntities(getFRUCB);
+
+  blkAndOp([]() { return false; }, 100);
+  toQueryData(results);
+
+  rowsQueue_.clear();
+}
+
+void IPMIClient::getMCs(QueryData& results) {
+  WriteLock lock(busy_);
+  rowsQueue_.clear();
+
+  iterateMCs(iterateMCsCB);
+
+  blkAndOp([]() { return false; }, 100);
+  toQueryData(results);
+
+  rowsQueue_.clear();
+}
+
+void IPMIClient::start() {
+  while (!interrupted()) {
+    pauseMilli(120000);
+    if (interrupted()) {
+      return;
+    }
+
+    if (isUp()) {
+      TLOG << "Running IPMIClient cleanup routine";
+      cleanup();
+    }
+  }
+}
+
+void IPMIClient::stop() {
+  if (up_.load()) {
+    up_.store(false);
+  }
+
+  cleanup();
+}
+
+void IPMIClient::cleanup() {
+  WriteLock lock(busy_);
+  if (!rowsQueue_.empty()) {
+    TLOG << "IPMIClient cleaning up rowsQueue_ since it's not empty";
+    rowsQueue_.clear();
+  }
+}
+
+namespace tables {
```

Comment:
I'd say here is the good differentiation for what separates implementation/header code from the table implementation code.